### PR TITLE
fix(exec): confirm the entry script is pnpm before re-invoking it

### LINKED ISSUE #14888: runPnpmCli identifies the running pnpm by basename alone
### What

`runPnpmCli` in `pnpm11/exec/pnpm-cli-runner/src/index.ts` decides whether to re-invoke `process.argv[1]` by matching its basename against a fixed list:

```ts
const PNPM_ENTRY_SCRIPTS = new Set(['pnpm', 'pn', 'pnpm.cjs', 'pnpm.mjs'])
```

Any other entry script falls back to whichever `pnpm` is on `PATH`. A script that is not pnpm but happens to carry one of those four names is re-executed as if it were.

### Why the list exists

It cannot simply be dropped. `runPnpmCli` also runs inside processes that merely import pnpm's command packages: the Jest suites for `verifyDepsBeforeRun` and `pack-app` call the handlers in process, so `process.argv[1]` is `jest.js` and an unconditional re-invoke turns the child into `node jest.js install`. That was tried and reverted in https://github.com/pnpm/pnpm/pull/14839.

`pnpx` and `pnx` are excluded on purpose, since that entry rewrites `process.argv` to prepend `dlx`.

### The gap

Reaching the bad case needs a host that imports `@pnpm/engine.runtime.commands`, `@pnpm/exec.commands` or `@pnpm/releasing.commands`, calls one of their handlers, and is named exactly `pnpm`, `pn`, `pnpm.cjs` or `pnpm.mjs`. That is unlikely, which is why https://github.com/pnpm/pnpm/pull/14839 left it alone rather than growing in scope. It is still worth closing, because the list has to track install shapes to stay correct.

### Why the obvious fix is wrong

`process.argv[1]` keeps the name the process was launched through, not the realpath. The two supported install shapes therefore disagree about what it looks like:

| Install method | `.bin` entry | `process.argv[1]` basename |
| --- | --- | --- |
| `npm install pnpm` | symlink to `bin/pnpm.mjs` | `pn`, `pnpm` |
| `pnpm add pnpm` | cmd-shim script | `pnpm.mjs` |
| Corepack | explicit path | `pnpm.cjs` |

So walking up from `process.argv[1]` for an owning `package.json`, the natural way to check ownership, finds the wrong manifest for the npm shape:

```
argv1 as launched : <project>/node_modules/.bin/pn
  naive walk up   : {"file":"<project>/package.json","name":"<project>"}
realpath          : <project>/node_modules/pnpm/bin/pnpm.mjs
  walk up         : {"file":"<project>/node_modules/pnpm/package.json","name":"pnpm"}
```

The walk lands on the consuming project, concludes the entry is not pnpm, and falls back to the `PATH` pnpm, which is the behavior https://github.com/pnpm/pnpm/pull/14839 set out to fix.

### Suggested shape

- Resolve `process.argv[1]` with `fs.realpathSync` before any ownership test.
- Walk up for the nearest `package.json` and accept `name` of `pnpm` or `@pnpm/exe`.
- Keep the basename set as a cheap pre-filter, so the common path stays free of filesystem work and `pnpx`/`pnx` keep their explicit exclusion.
- Fail **open** on any filesystem error, meaning re-invoke the entry. Failing closed silently restores the `PATH` fallback this is meant to remove.
- Still execute `process.argv[1]` rather than the realpath, so `--preserve-symlinks` setups are unaffected.

`resolveSelfCommand` in `pnpm11/pnpm/src/cmd/doctor.ts` and `getExecPath` in `pnpm11/engine/pm/commands/src/setup/setup.ts` answer the same question with no check at all. If this lands, a single shared helper for "the command that re-invokes the pnpm running now" would cover all three.

Raised by CodeRabbit on https://github.com/pnpm/pnpm/pull/14839.

---
Written by an agent (Claude Code, claude-opus-5).

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Closes pnpm/pnpm#14888.

`runPnpmCli` decided whether to re-invoke `process.argv[1]` from its basename alone, against a fixed set of pnpm's bin names. A script that merely carried one of those names was handed the command as if it were pnpm.

The set cannot be dropped, which pnpm/pnpm#14839 established the hard way: `runPnpmCli` also runs inside processes that only import pnpm's command packages, so an unconditional re-invoke turned a Jest host into `node jest.js install`. `pnpx` and `pnx` have to stay out for the mirror reason, since that entry rewrites `process.argv` to prepend `dlx`. So the names remain as a cheap pre-filter and the package manifest above the script now decides.

Three details are worth calling out for review.

**The check resolves symlinks first.** `process.argv[1]` keeps the name the process was launched through rather than the realpath, and the two install shapes disagree about what that looks like:

| Install method | `.bin` entry | `process.argv[1]` basename |
| --- | --- | --- |
| `npm install pnpm` | symlink to `bin/pnpm.mjs` | `pn`, `pnpm` |
| `pnpm add pnpm` | cmd-shim script | `pnpm.mjs` |
| Corepack | explicit path | `pnpm.cjs` |

Walking up from `process.argv[1]` without resolving finds the *consuming* project's manifest for the npm shape, concludes the entry is not pnpm, and falls back to the PATH lookup this exists to avoid:

```
argv1 as launched : <project>/node_modules/.bin/pn
  naive walk up   : {"file":"<project>/package.json","name":"<project>"}
realpath          : <project>/node_modules/pnpm/bin/pnpm.mjs
  walk up         : {"file":"<project>/node_modules/pnpm/package.json","name":"pnpm"}
```

The script that gets executed is still `process.argv[1]`, not the resolved path, so `--preserve-symlinks` setups are unaffected.

**Every filesystem failure leaves the script trusted.** Only a manifest that names some other package rejects it. An entry with no manifest above it, a copied bundle among them, is still pnpm as far as anything here can tell, and refusing it would put back the PATH lookup. Failing closed would silently undo pnpm/pnpm#14839.

**`pnpm doctor` shared the question and had no check at all.** `resolveSelfCommand` there re-invoked any `process.argv[1]`. It now uses the same `resolvePnpmSelfCommand`. Its fallback changes with that: a process with no usable entry script gets the pnpm on PATH rather than a bare `node`, which was never a runnable command. `pnpm setup`'s `getExecPath` is deliberately left alone, because it asks where this executable lives so it can install from that path, not what to run.

### Verification

Unit tests build real directory trees rather than asserting on mocks, since the check reads the filesystem. Removing the ownership check fails the new case and nothing else. Beyond the suite, every real entry point in this tree resolves as intended:

```
REUSED   pnpm11/pnpm/bin/pnpm.mjs
REUSED   pnpm11/pnpm/bin/pnpm.cjs
REUSED   pnpm11/pnpm/dist/pnpm.mjs
PATH     pnpm11/pnpm/bin/pnpx.mjs
REUSED   npm-installed .bin/pn
REUSED   npm-installed .bin/pnpm
PATH     npm-installed .bin/pnpx
PATH     npm-installed .bin/pnx
```

### No changeset

Deliberate, and easy to reverse if you disagree. The basename set this hardens landed in pnpm/pnpm#14839, whose changeset is still pending in `.changeset`, so it has not shipped. Nothing here changes what a user running pnpm normally observes: every real entry point resolves exactly as before, and only an embedded host that is not pnpm behaves differently.

## Squash Commit Body

```text
`runPnpmCli` decided whether to re-invoke `process.argv[1]` from its
basename alone, so a script that merely carried one of pnpm's bin names
was run as if it were pnpm. The names cannot be dropped -- a Jest host's
`process.argv[1]` must not be re-executed, and `pnpx` must not be, since
that entry rewrites `process.argv` to prepend `dlx` -- so the basename
set stays as a pre-filter and the manifest above the script now decides.

The check resolves symlinks first. `process.argv[1]` keeps the name the
process was launched through, so an npm-installed pnpm arrives as
`node_modules/.bin/pn`, whose nearest manifest is the consuming
project's rather than pnpm's. Walking up without resolving would reject
a real pnpm and fall back to the PATH lookup this exists to avoid.

Only a manifest that names some other package rejects the script. An
entry with no manifest above it, a copied bundle among them, is still
pnpm as far as anything here can tell, so every filesystem failure
leaves the script trusted.

`pnpm doctor` asked the same question -- what re-invokes the pnpm
running now -- and answered it with no check at all. It now shares
`resolvePnpmSelfCommand`. Its fallback changes with that: a process
with no usable entry script gets the pnpm on PATH rather than a bare
`node`, which was never a runnable command.

`pnpm setup`'s `getExecPath` is left alone. It asks where this
executable lives so it can install from that path, not what to run.

pnpm v12 resolves its own path with `std::env::current_exe()` and has
no name-based guess to correct.

Closes pnpm/pnpm#14888
```

## Checklist

- [x] I checked the referenced issue and verified that none of the PRs
  already linked to it solves it.
- [x] New features are implemented only in the Rust pnpm v12 CLI. Bug fixes
  are implemented in every affected version. pnpm v12 resolves its own path
  with `std::env::current_exe()`, so it has no name-based guess to correct.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Omitted on purpose, see "No changeset" above.
- [x] Added or updated tests.
- [ ] Updated the documentation if needed.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01W8gnb9oM214YsjkR4rouF5

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/14889"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1791977198&installation_model_id=426870&pr_number=14889&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F14889&signature=bb38eb9d2b30fd2e8290c54effce380674b4131437758ae8c3d2dbc2e717c9bb"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved pnpm executable detection for commands and diagnostics.
  - Prevented unrelated packages with pnpm-like binary names from being mistaken for pnpm.
  - Prevented recursive pnpm commands from executing package-installed scripts named `pnpm` or `pn`.
  - Added safer handling for missing, broken, malformed, copied, or symlinked package manifests.
  - Improved fallback behavior when entry-script resolution fails.

- **Refactor**
  - Standardized pnpm self-command resolution across command execution and diagnostics.

### CODE REVIEW

File: pnpm11/exec/pnpm-cli-runner/src/index.ts [hunk 1]
```diff
@@ -10,35 +11,86 @@ import { sync as execSync } from 'execa'
...
+ */
+function isPnpmEntryScript (entryScript: string): boolean {
+  if (!PNPM_ENTRY_SCRIPTS.has(path.basename(entryScript).toLowerCase())) return false
+  const owner = readOwningPackageName(entryScript)
+  // Only a package that identifies itself as something other than pnpm is
+  // rejected. An entry with no manifest above it, a copied bundle among them,
+  // is still pnpm as far as anything here can tell, and refusing it would put
+  // back the `PATH` lookup this exists to avoid.
+  return owner == null || PNPM_PACKAGE_NAMES.has(owner)
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/src/index.ts:80]: Good catch, and it was a regression against `main`: a copied bundle is accepted there today, and this PR would have started sending it to PATH. That is the behavior the PR exists to remove, so the rule was wrong.

Fixed in 3c7eeef22c by asking a narrower question. The owning package name no longer decides on its own; a script is rejected only when that package declares this exact file as one of its own bins. A third party publishing a `pn` bin does that. A bundle copied into a project does not, so it stays trusted.

Covered both forms of `bin`, the map and the string, since a package whose only bin is named after itself is the likely collision. There is a regression test for the copied bundle at `<project>/tools/pnpm.mjs`.

File: pnpm11/exec/pnpm-cli-runner/test/index.test.ts [hunk 2]
```diff
@@ -13,13 +17,38 @@ jest.unstable_mockModule('@pnpm/cli.meta', () => ({
...
+  while (tempDirs.length > 0) {
+    fs.rmSync(tempDirs.pop()!, { recursive: true, force: true })
+  }
 })
 
+/**
+ * An installed package whose bin is reached through a `node_modules/.bin` link,
+ * the shape npm creates. Returns the link, which is what `process.argv[1]`
+ * would be.
+ */
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/test/index.test.ts]: Trimmed in 3c7eeef22c. The helper now carries one line for the part its signature does not say, which is that the returned path is the `.bin` link rather than the script, because that is what `process.argv[1]` would be.

File: pnpm11/exec/pnpm-cli-runner/test/index.test.ts [hunk 3]
```diff
@@ -32,12 +61,54 @@ test('the @pnpm/exe build re-runs itself', () => {
...
-  '/path/to/pnpm/bin/pnpm.cjs',
-  '/path/to/pnpm/dist/pnpm.mjs',
-  '/path/to/node_modules/.bin/pnpm',
-  '/path/to/node_modules/.bin/pn',
-])('the entry script %s is re-run with Node.js', (entryScript) => {
+  ['pnpm', 'pnpm.mjs'],
+  ['pn', 'pnpm.mjs'],
+  ['pnpm', 'pnpm.cjs'],
+])('pnpm reached through node_modules/.bin/%s is re-run with Node.js', (binName, scriptName) => {
+  const entryScript = installPackage('pnpm', binName, scriptName)
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/test/index.test.ts]: Right, and it was load-bearing. Added in e1e2147353. Narrowing `PNPM_PACKAGE_NAMES` to just `pnpm` now fails that case and only that case, so the suite no longer passes an implementation that ignores the second name.

File: pnpm11/pnpm/src/cmd/doctor.ts [hunk 4]
```diff
@@ -93,7 +94,7 @@ export interface DoctorCommandOptions {
 const DEFAULT_REGISTRY = 'https://registry.npmjs.org/'
 
 export async function handler (opts: DoctorCommandOptions): Promise<{ output: string, exitCode: number }> {
-  const pnpmCommand = opts.pnpmCommand ?? resolveSelfCommand()
+  const pnpmCommand = opts.pnpmCommand ?? resolvePnpmSelfCommand()
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/pnpm/src/cmd/doctor.ts:97]: Added in 3c7eeef22c.

I had left it out on the grounds that the set of names being hardened came from pnpm/pnpm#14839 and has not shipped. That was too narrow a reading. Released pnpm re-invokes `process.argv[1]` on a basename match too, so the entry check applies to shipped behavior and the exemption does not.

**Comment by zkochan** [comment]
Addressing the two Qodo findings that are still open. Neither has an inline thread, so they are here.

**3, entry-check docs overstate trust.** Fair against the text it read, and the rewrite in 3c7eeef22c changed it for a different reason. The doc now states the weak guarantee rather than implying a strong one:

> Whether `entryScript` can be re-invoked as pnpm. The name alone is not enough, since these are ordinary enough for another package to publish as a bin, so a script carrying one is rejected when the package owning it claims it as a bin of its own.
>
> Nothing stronger is asked of it. pnpm is distributed as a copyable bundle and reached through shims and links, so a script that no package claims is pnpm as far as anything here can tell, and refusing it would put back the `PATH` lookup this exists to avoid.

The fail-open behavior is now the stated contract, not an unmentioned branch.

**2, symlink swaps can execute another script.** Declining, and leaving it visible rather than resolving it, in case a maintainer weighs it differently.

The window is between `realpathSync` and the spawn, and reaching it needs write access to the entry path or the `.bin` directory holding it. Anyone with that access can replace the target script itself, or any other bin in that directory the user runs, without racing anything. So pinning execution to the validated path does not remove the capability, it only removes one of several equivalent routes to it.

The suggested fix also has a real cost. Executing the resolved path instead of `process.argv[1]` changes what the child sees as its own entry, which is exactly what pnpm/pnpm#14797 went to some length to keep faithful in the alias bins, and it changes behavior under `--preserve-symlinks`. That is a wider blast radius than the race it closes.

Worth noting on scope: released pnpm re-invokes `process.argv[1]` with no ownership check at all. This PR narrows what gets re-invoked; it does not widen it.

---
Written by an agent (Claude Code, claude-opus-5).

File: pnpm11/exec/pnpm-cli-runner/src/index.ts [hunk 5]
```diff
@@ -10,35 +11,102 @@ import { sync as execSync } from 'execa'
...
+    try {
+      contents = fs.readFileSync(path.join(dir, 'package.json'), 'utf8')
+    } catch {
+      const parent = path.dirname(dir)
+      if (parent === dir) return undefined
+      dir = parent
+      continue
+    }
+    try {
+      return { dir, manifest: JSON.parse(contents) as OwningPackage['manifest'] }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/src/index.ts]: Confirmed and fixed in 29eaac01c9. Reproduced first:

```
THREW: TypeError Cannot read properties of null (reading 'name')
```

`readOwningPackage` now rejects any parse result that is not a non-null object, so a manifest holding `null` joins the unreadable ones and the entry stays trusted instead of taking the command down. Swept the shapes that parse to a non-object, all of which now fall back rather than throw:

```
"null"         REUSED
"\"a string\"" REUSED
"123"          REUSED
"[]"           REUSED
"{ not json"   REUSED
""             REUSED
```

Regression test covers `null` alongside the existing malformed-JSON case.

---
Written by an agent (Claude Code, claude-opus-5).

**Comment by zkochan** [comment]
Two Qodo findings from the latest pass, neither actioned. Reasoning below so a maintainer can overrule either.

**5, entry-check docs overstate trust.** Stale. It quotes `readOwningPackageName` and the line "The name is only a hint, so confirm the script is pnpm's before handing it", both of which were replaced in 3c7eeef22c. The current doc states the fail-open contract outright rather than implying verification:

> Nothing stronger is asked of it. pnpm is distributed as a copyable bundle and reached through shims and links, so a script that no package claims is pnpm as far as anything here can tell, and refusing it would put back the `PATH` lookup this exists to avoid.

The behavior it describes, failing open on a missing file or an unreadable manifest, is deliberate and now tested: the `null`, malformed-JSON, broken-symlink and no-manifest cases all assert the entry stays trusted. Failing closed there would silently restore the `PATH` fallback that pnpm/pnpm#14839 removed.

**2, test helper docs hide disk writes.** Declining. It asks `installPackage` to document its "name inputs, created package tree and symlink, cleanup registration, and filesystem failure behavior", which is a description of the fifteen lines directly beneath it. CLAUDE.md is explicit that this is not wanted:

> **Do not write a comment** that restates what the code already says.
>
> **Documentation comments are for the item's contract** [...] Not for re-narrating the body.

Worth noting the two bots want opposite things here. Greptile asked for this same comment to be deleted as narration; the one line that survived documents the only part the signature does not give away, which is that the returned path is the `.bin` link rather than the script, because that is what `process.argv[1]` would be. I think that is the right amount, but it is a small call and easy to change either way.

**4, symlink swaps**, answered in https://github.com/pnpm/pnpm/pull/14889#issuecomment-5663299355 and left open on purpose.

---
Written by an agent (Claude Code, claude-opus-5).

File: pnpm11/exec/pnpm-cli-runner/src/index.ts [hunk 6]
```diff
@@ -10,35 +11,102 @@ import { sync as execSync } from 'execa'
...
+    } catch {
+      return undefined
+    }
+  }
+}
+
+function declaresBin ({ dir, manifest }: OwningPackage, resolvedScript: string): boolean {
+  const { bin } = manifest
+  const targets = typeof bin === 'string' ? [bin] : typeof bin === 'object' && bin !== null ? Object.values(bin) : []
+  return targets.some((target) => typeof target === 'string' && path.resolve(dir, target) === resolvedScript)
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/src/index.ts]: Real gap, fixed in 9e6c6c2ded. `declaresBin` compared a resolved script against unresolved manifest paths, so a package pointing a bin at a symlink inside its own tree never matched and its script was re-invoked as pnpm.

Declared targets are now resolved the same way the entry script is, falling back to the literal path when a target cannot be resolved, so a declared bin that does not exist rejects nothing.

The regression test builds the full chain, `.bin/pn` to `lib/pn` to `cli.js`, plus a missing first bin entry so the failing resolve is exercised on the way to the matching one. Comparing unresolved paths again fails that case and only that case.

---
Written by an agent (Claude Code, claude-opus-5).

File: pnpm11/exec/pnpm-cli-runner/test/index.test.ts [hunk 7]
```diff
@@ -32,33 +40,120 @@ test('the @pnpm/exe build re-runs itself', () => {
...
+  fs.symlinkSync(path.join(root, 'missing.mjs'), entryScript)
+  process.argv[1] = entryScript
+
+  expectReinvoked(entryScript)
+})
+
+test.each([
+  '{ not json',
+  'null',
+])('an entry script under the manifest %s is re-run', (contents) => {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/exec/pnpm-cli-runner/test/index.test.ts:90]: Added in 948ce5d184, with `[]` as well since an array is another valid non-object shape.

The reasoning needs one correction, though. These rows do not fail against a guard that rejects only `null`. I checked by narrowing the guard to exactly that and rerunning, and all 21 tests still passed. The other shapes never violate the contract in the first place:

```
null     THROWS on field access
"pnpm"   name=undefined bin=undefined
1        name=undefined bin=undefined
[]       name=undefined bin=undefined
```

A bare string, a number and an array carry no `name` and no `bin`, so they already fail open by the ordinary route and behave identically either way. `null` is the only shape that throws, which is why it was the only bug.

So the new rows document the fail-open contract rather than guarding a regression. Worth keeping for that, and the commit message says as much, but I did not want to claim coverage they do not provide.

---
Written by an agent (Claude Code, claude-opus-5).
